doc/source/actions.rst

```diff
@@ -351,6 +351,26 @@ The ``merge`` action merges the pull request into its base branch. The
        Note that the ``rebase`` method has many drawbacks, see :ref:`strict
        rebase`.
 
+   * - ``strict_smart_behavior``
+     - string
+     - ``merge-if-ready``
+     - Behavior when a pull request is ready to be merge in ``strict=smart``
+       mode. Possible values:
+
+       * ``merge-if-ready`` if the pull request is ready and already in sync
+         with its base branch, the queue is by-passed and the pull request is
+         directly merged. This behavior allows to safe CIs resources.
```

Comment:
If you have a long-running CIs, you really save resources.

Assume CI takes 2 hours:

merge-if-ready:
* PR 1 is ready except CIs CI already ran since 1 hour
* PR 2 just become ready and already up2date, we just merge it without queuing it.
* PR 1 get rebased, previous CI RUN get canceled, we wait 2h and  merged it

CIs total runs times 3 hours

always-append:
* PR 1 is ready except CIs, CI already run since 1 hour
* PR 2 just become ready and already up2date, we put it in queue.
* PR 1 CIs finish 1 hour later and get merged
* PR 2 is rebased, CIs rerun and take 2 hours
* PR 2 get merged

CIs total runs times: 4 hours

Reply:
I think you forgot some:

`merge-if-ready`:

- PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)
- PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)

CIs total runs times **5 hours**

`always-append`:

- PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
- PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
- PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
- PR 2 get merged

CIs total runs times: **5 hours**

So it's the same, no?

Reply:
> I think you forgot some:
> 
> `merge-if-ready`:
> 
>     * PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)

PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

>     * PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)
> 
> 
> CIs total runs times **5 hours**
> 
> `always-append`:
> 
>     * PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
>     * PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
> 
>     * PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
> 
>     * PR 2 get merged
> 
> 
> CIs total runs times: **5 hours**
> 
> So it's the same, no?


Another way to explain it:

In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).
But the counter-part is that the first PR of the queue get its CIs canceled after 1 hour of run, rebased again and then 2 more hours of CIs , so 5 hours in total (2 hours before "merge" action is triggered, 1 hour for nothing, 2 hours after the rebased)
Here, for the two PRs, we have 7 hours in total.

Reply:
> PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

Of course, you have: for the CI to be green, it has to run for 2 hours. :)

> In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

👍  that's a good summary. In `always-append`, you always have 2 CI runs; the _election_ CI runs, where you make sure it works and it can enter the queue, and the _merge_ CI run that is done once it's at the top of the queue and going to be merged.

> In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).

The thing is that the more you cancel, the more CI time you use for nothing.
In `always-append` the CI time is constant and always _election_+_merge_. In `merge-if-ready`, you always have _election_, but then your _merge_ can be between 0 (the _election_ run is used) or infinite (you keep being canceled).

So it *can* save CI time, but it can also consume CI time forever 😨 😆